### ISSUE TITLE
pubsub example vs current data

### ISSUE THREAD

**Issue by kevmoo** [googleapis/google-cloudevents#151]
I *just* deployed a handler for pubsub using eventarc.

Here's the data I got:

```json
{
   "message":{
      "data":"RG8gdGhhdCB0aGluZywgeW8hIEkgbG92ZSBpdA==",
      "messageId":"1981947010177954",
      "message_id":"1981947010177954",
      "publishTime":"2021-02-05T04:06:14.109Z",
      "publish_time":"2021-02-05T04:06:14.109Z"
   },
   "subscription":"projects/kevmoo-firebase3-test/subscriptions/eventarc-us-central1-events-pubsub-trigger-sub-072"
}
```

VS

https://github.com/googleapis/google-cloudevents/blob/fc742ace41f7854ed67e3314af45d7e5c5260654/testdata/google/events/cloud/pubsub/v1/MessagePublishedData-text.json#L1-L10

Where is the existing behavior documented?

Here are the request headers as an fyi

```json
{
   "accept-encoding":"gzip,deflate,br",
   "accept":"application/json",
   "ce-id":"1981947010177954",
   "ce-source":"//pubsub.googleapis.com/projects/[redacted]/topics/eventarc-us-central1-events-pubsub-trigger-072",
   "ce-specversion":"1.0",
   "ce-time":"2021-02-05T04:06:14.109Z",
   "ce-type":"google.cloud.pubsub.topic.v1.messagePublished",
   "content-length":"323",
   "content-type":"application/json",
   "forwarded":"for=\"66.102.6.91\";proto=https",
   "from":"noreply@google.com",
   "host":"helloworld-events-rgvedcg7vq-uc.a.run.app",
   "user-agent":"APIs-Google; (+https://developers.google.com/webmasters/APIs-Google.html)",
   "x-cloud-trace-context":"fd349b66cc9a95bcea6ae4769c9c51fb/9609727294171356041;o=1",
   "x-forwarded-for":"66.102.6.91",
   "x-forwarded-proto":"https"
}
```

**Comment by kevmoo** [comment]
tl;dr: Should the messageId and publshTime be documented? Etc etc

**Comment by grant** [comment]
What's the exact question? Sorry, can you be clearer?

Are you saying the testdata file is incorrect? Are you looking for more test data?

- Attributes are optional – you need to add a attr with the pubsub message for that field to appear
- Eventarc add extra like `@type` that isn't part of the CloudEvent

**Comment by jskeet** [comment]
I think there are two different aspects here:

- Should a message have a message ID and a publish time? The message ID part doesn't surprise me at all, given that it's in [the proto](https://github.com/googleapis/google-cloudevents/blob/master/proto/google/events/cloud/pubsub/v1/data.proto). The publish time is something we should add to the proto if we expect it to continue to be present.
- Should a message have those attributes *twice* using different conventions?  (`publishTime` vs `publish_time`, and `messageId` vs `message_id`) That's more of a surprise to me, and we'll need to follow that up.

**Comment by kevmoo** [comment]
@jskeet thanks for interpreting my 🤷  of a issue. The `@type` bit is also inconsistant.

I'm working on the Dart impl here – having ~tight alignment between the docs and the shipping behavior would be goodness.

**Comment by jskeet** [comment]
I can at least *explain* the presence of `@type` - it occurs when you create a JSON representation of the `google.protobuf.Any` type, which is essentially "here's the protobuf binary data, along with a type URL to show the message it represents".
You can safely ignore it.

But I'd definitely make sure that your Dart implementation can handle the presence of fields that aren't currently in the .proto files - because new fields can be introduced at any time.

**Comment by kevmoo** [comment]
We just ignore fields we don't know about.

...just want to make sure we wire against the right fields and don't miss
anything.

Like, which messageId and date fields to use. :-)

**Comment by grant** [comment]
See b/164484479 for the duplicate messageId and date fields.

**Comment by grant** [comment]
We need docs to state this when using Eventarc.

**Comment by kevmoo** [comment]
Thanks! Could we update the test data, too? @jskeet ?

**Comment by jskeet** [comment]
@kevmoo: Maybe. At the moment it really depends on where you're consuming the data. I don't *think* I'd include the legacy fields in the test data, but we'd want to add the publish time.

**Comment by kevmoo** [comment]
I'm consuming the data from `pubsub` via eventarc. See the top of the issue to see the data I'm getting.

Just wanting what I'm seeing to be documented somewhere.

**Comment by jskeet** [comment]
Yes - my point is that if I change the test data to show the content received in GCF or other environments at the moment, then we could end up with another issue about that.

The point of adding the note in the now-merged PR is so that you shouldn't *expect* to see every field that might be present in reality in the test data. I wouldn't want to add the snake_case fields for example. (I may or may not remove the `@type` field from the current test data.)

It would be good for the test data to show all the fields you should reasonably rely on, but where there are other fields present for legacy or other reasons, I'd prefer not to show those in the test data.

**Comment by kevmoo** [comment]
Sure. But publishTime and messageId would be great...

Otherwise I'm not sure what I'm supposed to be parsing or not.

**Comment by jskeet** [comment]
`messageId` is already present in https://github.com/googleapis/google-cloudevents/blob/master/testdata/google/events/cloud/pubsub/v1/MessagePublishedData-binary.json

We could potentially add it to the text one as well. As as I say, I'll add `publishTime`.

But in general, if you see a field in reality and it's in the .proto file, then yes, you should absolutely treat that as definitive. (Note that the protos haven't been declared stable yet, but they will be eventually...) Note that just because a field *may* be there, and interpreted in a particular way if so, doesn't mean that it *will* be present in every message. We don't yet make any guarantees about particular fields always being populated. We *may* do that in the future, but only when we can be confident in those guarantees at a platform level.
